### PR TITLE
Action queue redo

### PR DESCRIPTION
**Pull request by grilledham**
I'm making this pull request so that I can reference it in #157

### CODE REVIEW

File: map_layout.lua [hunk 1]
```diff
@@ -6,7 +6,7 @@ in this file and your run_*type*_module(event) function will be called.
 --]]
 
 --combined--
---require "map_gen.combined.island_resort"
+require "map_gen.combined.island_resort"
```
Comment (Valansch) [inline map_layout.lua]: :D

**Comment by Valansch** [comment]
Why did you rename the callbacks to actions? Didn't we want to name both "task"?

**Comment by grilledham** [comment]
I wrote this before I saw that you agreed with my renaming suggestion. I didn't want to presume so I kept the action name. 

I'll make the renaming changes tomorrow.

**Comment by grilledham** [comment]
Change the task for island_resort to reusable tasks.

I used to run the map at 90 - 110 ups, now I can run it at 160 - 175 ups.

Tested on the test server for desyncs.

**Comment by grilledham** [comment]
90 - 110 is with the new queue, 160-175 is with reusable tasks. I think it was around 20 - 30 with the old queue.
My guess is the extra performance comes from less data being serialized, whenever the global table gets serialized. Could also be less pressure on the garbage collector.

**Comment by grilledham** [comment]
Ok I ran the test again, this time multiplying the queue size by 34 in the get_task_per_tick function.
I get around 110 - 120 ups now.

**Comment by Valansch** [comment]
>My guess is the extra performance comes from less data being serialized, whenever the global table gets serialized. Could also be less pressure on the garbage collector.

Agree thats most likely all contributing to better UPS.

>Or it could be that with a smaller queue, it is running less tasks per tick. In which case it's hard to measure the performance impact.

That shouldn't matter, if game speed is at some very high value. As the game is already running as fast as possible.

**Comment by Valansch** [comment]
>Ok, I ran the test again, this time multiplying the queue size by 34 in the get_task_per_tick function.
I get around 110 - 120 ups now.

Do you get different performance simply by multiplying the queue size by a value?

**Comment by grilledham** [comment]
Yes, because it's running more tasks per tick

**Comment by grilledham** [comment]
2 major changes:
1. Changed callbacks to use a priority queue, this simplifies the code in Task.lua.
2. Callbacks now store a string for the function name rather than the function itself, this makes them consistent with how tasks work.

1 minor change - Added global.task_queue_speed.

All tested for desyncs on the test server.

After your review/ approval I think this branch is ready to be merged.

**Comment by Valansch** [comment]
I think we can optimize get_task_per_tick a bit. We dont need it to change too often. Since its running every tick i dont want to caclulate log10 every tick.

```lua
global.tpt = global.task_queue_speed
function get_task_per_tick()  
  if game.tick % 300 == 0 then
    local size = global.total_task_weight
    local tpt = math.floor(math.log10(size + 1)) * global.task_queue_speed
    if apt < 1 then
      global.tpt = tpt
    else
      global.tpt = tpt
    end
  end
  return global.tpt
end
```

**Comment by Valansch** [comment]
I dont understand what the purpose of the PriorityQueue is. Whats the difference and why couldn't you use the normal Queue?

**Comment by grilledham** [comment]
A Priority Queue is like a normal Queue except the 'highest' priority element is removed each time pop is called. In this case the 'highest' priority element is the callback with the lowest time value.

A normal queue can't be used as there is no guarantee that the order in which the callbacks are inserted is the same as the order in which they should be removed. For example this code.
```
Task.set_timeout(20, "some_function")
Task.set_timeout(10, "some_function")
```

**Comment by Valansch** [comment]
Okay that makes sense. This looks pretty good. I see nothing more than my comment earlier that i would change. Want me to merge, or do you want to change that?

**Comment by Valansch** [comment]
I'll merge now and test.

**Comment by grilledham** [comment]
Sorry I was just taking a long time to think about your optimisation proposal for get_task_per_tick().
Without benchmarking the performance impact, I'm wary of making any micro optimisations that make the code more complicated. 

Optimising code is still fun though, so perhaps this is an issues we can revisit in the future.

**Comment by Valansch** [comment]
Nothing to be sorry about. I only waited an hour :D I wanted to launch the map, so i merged. I also applied the fix. Now its too late :P Unless you want to reverse it. I don't think it makes it too complicated.

**Comment by grilledham** [comment]
Merging was the right thing to do. The code that is in this branch was ready and that's all that matters. Regarding your fix, I don't think it does any harm, so I'm happy with it.

**Pull request merged**